File: sdk/python/xai_sdk/clients/transaction_client.py
"""
Transaction Client for XAI SDK

Handles all transaction-related operations.
"""

from typing import Optional, Dict, Any
from datetime import datetime

from ..http_client import HTTPClient
from ..models import Transaction, TransactionStatus
from ..exceptions import TransactionError, ValidationError


class TransactionClient:
    """Client for transaction operations."""

    def __init__(self, http_client: HTTPClient) -> None:
        """
        Initialize Transaction Client.

        Args:
            http_client: HTTP client instance
        """
        self.http_client = http_client

    def send(
        self,
        from_address: str,
        to_address: str,
        amount: str,
        data: Optional[str] = None,
        gas_limit: Optional[str] = None,
        gas_price: Optional[str] = None,
        nonce: Optional[int] = None,
        signature: Optional[str] = None,
    ) -> Transaction:
        """
        Send a transaction.

        Args:
            from_address: Sender address
            to_address: Recipient address
            amount: Transaction amount
            data: Optional transaction data
            gas_limit: Gas limit for transaction
            gas_price: Gas price per unit
            nonce: Transaction nonce
            signature: Signed transaction data

        Returns:
            Transaction details

        Raises:
            TransactionError: If transaction fails
        """
        if not from_address or not to_address or not amount:
            raise ValidationError("from_address, to_address, and amount are required")

        try:
            payload = {
                "from": from_address,
                "to": to_address,
                "amount": amount,
            }

            if data:
                payload["data"] = data
            if gas_limit:
                payload["gas_limit"] = gas_limit
            if gas_price:
                payload["gas_price"] = gas_price
            if nonce is not None:
                payload["nonce"] = nonce
            if signature:
                payload["signature"] = signature

            response = self.http_client.post("/transaction/send", data=payload)

            return Transaction(
                hash=response["hash"],
                from_address=response["from"],
                to_address=response["to"],
                amount=response["amount"],
                timestamp=datetime.fromisoformat(response["timestamp"]),
                status=TransactionStatus(response.get("status", "pending")),
                fee=response.get("fee", "0"),
                gas_used=response.get("gas_used", "0"),
            )
        except TransactionError:
            raise
        except Exception as e:
            raise TransactionError(f"Failed to send transaction: {str(e)}")

    def get(self, tx_hash: str) -> Transaction:
        """
        Get transaction details.

        Args:
            tx_hash: Transaction hash

        Returns:
            Transaction details

        Raises:
            TransactionError: If transaction retrieval fails
        """
        if not tx_hash:
            raise ValidationError("tx_hash is required")

        try:
            response = self.http_client.get(f"/transaction/{tx_hash}")

            return Transaction(
                hash=response["hash"],
                from_address=response["from"],
                to_address=response["to"],
                amount=response["amount"],
                timestamp=datetime.fromisoformat(response["timestamp"]),
                status=TransactionStatus(response.get("status", "pending")),
                fee=response.get("fee", "0"),
                gas_used=response.get("gas_used", "0"),
                block_number=response.get("block_number"),
                block_hash=response.get("block_hash"),
                confirmations=response.get("confirmations", 0),
            )
        except Exception as e:
            raise TransactionError(f"Failed to get transaction: {str(e)}")

    def get_status(self, tx_hash: str) -> Dict[str, Any]:
        """
        Get transaction status.

        Args:
            tx_hash: Transaction hash

        Returns:
            Transaction status information

        Raises:
            TransactionError: If status retrieval fails
        """
        if not tx_hash:
            raise ValidationError("tx_hash is required")

        try:
            return self.http_client.get(f"/transaction/{tx_hash}/status")
        except Exception as e:
            raise TransactionError(f"Failed to get transaction status: {str(e)}")

    def estimate_fee(
        self,
        from_address: str,
        to_address: str,
        amount: str,
        data: Optional[str] = None,
    ) -> Dict[str, Any]:
        """
        Estimate transaction fee.

        Args:
            from_address: Sender address
            to_address: Recipient address
            amount: Transaction amount
            data: Optional transaction data

        Returns:
            Fee estimation with gas details

        Raises:
            TransactionError: If estimation fails
        """
        if not from_address or not to_address or not amount:
            raise ValidationError("from_address, to_address, and amount are required")

        try:
            payload = {
                "from": from_address,
                "to": to_address,
                "amount": amount,
            }
            if data:
                payload["data"] = data

            return self.http_client.post("/transaction/estimate-fee", data=payload)
        except Exception as e:
            raise TransactionError(f"Failed to estimate fee: {str(e)}")

    def is_confirmed(self, tx_hash: str, confirmations: int = 1) -> bool:
        """
        Check if transaction is confirmed.

        Args:
            tx_hash: Transaction hash
            confirmations: Number of required confirmations

        Returns:
            True if transaction is confirmed

        Raises:
            TransactionError: If check fails
        """
        try:
            status = self.get_status(tx_hash)
            return status.get("confirmations", 0) >= confirmations
        except Exception as e:
            raise TransactionError(f"Failed to check confirmation: {str(e)}")

    def wait_for_confirmation(
        self,
        tx_hash: str,
        confirmations: int = 1,
        timeout: int = 600,
        poll_interval: int = 5,
    ) -> Transaction:
        """
        Wait for transaction confirmation.

        Args:
            tx_hash: Transaction hash
            confirmations: Number of required confirmations
            timeout: Maximum time to wait in seconds
            poll_interval: Polling interval in seconds

        Returns:
            Confirmed transaction

        Raises:
            TransactionError: If confirmation fails or times out
        """
        import time

        start_time = time.time()

        while True:
            if time.time() - start_time > timeout:
                raise TransactionError(
                    f"Transaction confirmation timeout after {timeout}s"
                )

            try:
                tx = self.get(tx_hash)
                if tx.confirmations >= confirmations:
                    return tx
            except Exception as e:
                raise TransactionError(f"Failed to wait for confirmation: {str(e)}")

            time.sleep(poll_interval)

File: src/xai/sdk/typescript/src/clients/transaction-client.ts
/**
 * Transaction Client for XAI SDK
 *
 * Handles transaction sending, retrieval, fee estimation, and confirmation tracking.
 */

import { HTTPClient } from '../utils/http-client';
import {
  Transaction,
  TransactionStatus,
  SendTransactionParams,
  FeeEstimation,
} from '../types';
import { TransactionError, ValidationError } from '../errors';

/**
 * Client for transaction operations
 */
export class TransactionClient {
  constructor(private httpClient: HTTPClient) {}

  /**
   * Send a transaction
   *
   * @param params - Transaction parameters
   * @returns Transaction details
   *
   * @example
   * ```typescript
   * const tx = await client.transaction.send({
   *   from: '0x1234...',
   *   to: '0x5678...',
   *   amount: '1000',
   *   data: '0x...'
   * });
   * console.log('Transaction hash:', tx.hash);
   * ```
   */
  async send(params: SendTransactionParams): Promise<Transaction> {
    if (!params.from || !params.to || !params.amount) {
      throw new ValidationError('from, to, and amount are required');
    }

    try {
      const payload: Record<string, unknown> = {
        from: params.from,
        to: params.to,
        amount: params.amount,
      };

      if (params.data) {
        payload.data = params.data;
      }
      if (params.gasLimit) {
        payload.gas_limit = params.gasLimit;
      }
      if (params.gasPrice) {
        payload.gas_price = params.gasPrice;
      }
      if (params.nonce !== undefined) {
        payload.nonce = params.nonce;
      }
      if (params.signature) {
        payload.signature = params.signature;
      }

      const response = await this.httpClient.post<{
        hash: string;
        from: string;
        to: string;
        amount: string;
        timestamp: string;
        status?: string;
        fee?: string;
        gas_used?: string;
      }>('/transaction/send', payload);

      return {
        hash: response.hash,
        from: response.from,
        to: response.to,
        amount: response.amount,
        timestamp: response.timestamp,
        status: (response.status as TransactionStatus) || TransactionStatus.PENDING,
        fee: response.fee || '0',
        gasUsed: response.gas_used || '0',
      };
    } catch (error) {
      if (error instanceof TransactionError) {
        throw error;
      }
      throw new TransactionError(`Failed to send transaction: ${error}`);
    }
  }

  /**
   * Get transaction details
   *
   * @param txHash - Transaction hash
   * @returns Transaction details
   *
   * @example
   * ```typescript
   * const tx = await client.transaction.get('0xabc...');
   * console.log('Transaction status:', tx.status);
   * console.log('Confirmations:', tx.confirmations);
   * ```
   */
  async get(txHash: string): Promise<Transaction> {
    if (!txHash) {
      throw new ValidationError('txHash is required');
    }

    try {
      const response = await this.httpClient.get<{
        hash: string;
        from: string;
        to: string;
        amount: string;
        timestamp: string;
        status?: string;
        fee?: string;
        gas_used?: string;
        block_number?: number;
        block_hash?: string;
        confirmations?: number;
      }>(`/transaction/${txHash}`);

      return {
        hash: response.hash,
        from: response.from,
        to: response.to,
        amount: response.amount,
        timestamp: response.timestamp,
        status: (response.status as TransactionStatus) || TransactionStatus.PENDING,
        fee: response.fee || '0',
        gasUsed: response.gas_used || '0',
        blockNumber: response.block_number,
        blockHash: response.block_hash,
        confirmations: response.confirmations || 0,
      };
    } catch (error) {
      if (error instanceof TransactionError) {
        throw error;
      }
      throw new TransactionError(`Failed to get transaction: ${error}`);
    }
  }

  /**
   * Get transaction status
   *
   * @param txHash - Transaction hash
   * @returns Transaction status information
   *
   * @example
   * ```typescript
   * const status = await client.transaction.getStatus('0xabc...');
   * console.log('Status:', status);
   * ```
   */
  async getStatus(txHash: string): Promise<Record<string, unknown>> {
    if (!txHash) {
      throw new ValidationError('txHash is required');
    }

    try {
      return await this.httpClient.get(`/transaction/${txHash}/status`);
    } catch (error) {
      if (error instanceof TransactionError) {
        throw error;
      }
      throw new TransactionError(`Failed to get transaction status: ${error}`);
    }
  }

  /**
   * Estimate transaction fee
   *
   * @param params - Transaction parameters for estimation
   * @returns Fee estimation details
   *
   * @example
   * ```typescript
   * const fee = await client.transaction.estimateFee({
   *   from: '0x1234...',
   *   to: '0x5678...',
   *   amount: '1000'
   * });
   * console.log('Estimated fee:', fee.estimatedFee);
   * console.log('Gas limit:', fee.gasLimit);
   * ```
   */
  async estimateFee(params: {
    from: string;
    to: string;
    amount: string;
    data?: string;
  }): Promise<FeeEstimation> {
    if (!params.from || !params.to || !params.amount) {
      throw new ValidationError('from, to, and amount are required');
    }

    try {
      const payload: Record<string, unknown> = {
        from: params.from,
        to: params.to,
        amount: params.amount,
      };

      if (params.data) {
        payload.data = params.data;
      }

      const response = await this.httpClient.post<{
        estimated_fee: string;
        gas_limit: string;
        gas_price: string;
        base_fee?: string;
        priority_fee?: string;
      }>('/transaction/estimate-fee', payload);

      return {
        estimatedFee: response.estimated_fee,
        gasLimit: response.gas_limit,
        gasPrice: response.gas_price,
        baseFee: response.base_fee,
        priorityFee: response.priority_fee,
      };
    } catch (error) {
      if (error instanceof TransactionError) {
        throw error;
      }
      throw new TransactionError(`Failed to estimate fee: ${error}`);
    }
  }

  /**
   * Check if transaction is confirmed
   *
   * @param txHash - Transaction hash
   * @param confirmations - Number of required confirmations (default: 1)
   * @returns True if transaction has required confirmations
   *
   * @example
   * ```typescript
   * const isConfirmed = await client.transaction.isConfirmed('0xabc...', 3);
   * if (isConfirmed) {
   *   console.log('Transaction confirmed with 3+ confirmations');
   * }
   * ```
   */
  async isConfirmed(txHash: string, confirmations: number = 1): Promise<boolean> {
    try {
      const status = await this.getStatus(txHash);
      return (status.confirmations as number) >= confirmations;
    } catch (error) {
      if (error instanceof TransactionError) {
        throw error;
      }
      throw new TransactionError(`Failed to check confirmation: ${error}`);
    }
  }

  /**
   * Wait for transaction confirmation
   *
   * @param txHash - Transaction hash
   * @param confirmations - Number of required confirmations (default: 1)
   * @param timeout - Maximum time to wait in milliseconds (default: 600000)
   * @param pollInterval - Polling interval in milliseconds (default: 5000)
   * @returns Confirmed transaction
   *
   * @example
   * ```typescript
   * const tx = await client.transaction.waitForConfirmation('0xabc...', 3);
   * console.log('Transaction confirmed!', tx);
   * ```
   */
  async waitForConfirmation(
    txHash: string,
    confirmations: number = 1,
    timeout: number = 600000,
    pollInterval: number = 5000
  ): Promise<Transaction> {
    const startTime = Date.now();

    while (true) {
      if (Date.now() - startTime > timeout) {
        throw new TransactionError(`Transaction confirmation timeout after ${timeout}ms`);
      }

      try {
        const tx = await this.get(txHash);
        if ((tx.confirmations || 0) >= confirmations) {
          return tx;
        }
      } catch (error) {
        if (error instanceof TransactionError) {
          throw error;
        }
        throw new TransactionError(`Failed to wait for confirmation: ${error}`);
      }

      await new Promise((resolve) => setTimeout(resolve, pollInterval));
    }
  }

  /**
   * Get pending transactions
   *
   * @returns List of pending transactions
   *
   * @example
   * ```typescript
   * const pending = await client.transaction.getPending();
   * console.log('Pending transactions:', pending);
   * ```
   */
  async getPending(): Promise<Transaction[]> {
    try {
      const response = await this.httpClient.get<{
        transactions: Transaction[];
      }>('/mempool');

      return response.transactions || [];
    } catch (error) {
      if (error instanceof TransactionError) {
        throw error;
      }
      throw new TransactionError(`Failed to get pending transactions: ${error}`);
    }
  }
}
